File: src/algorithms/strings/edit-distance/longest-common-substring/__tests__/longest-common-substring_test.go
package main

import "testing"

func TestLongestCommonSubstringAbabcBabcba(t *testing.T) {
	if longestCommonSubstring("ABABC", "BABCBA") != 4 {
		t.Error("expected 4")
	}
}

func TestLongestCommonSubstringSourceEmpty(t *testing.T) {
	if longestCommonSubstring("", "abc") != 0 {
		t.Error("expected 0 when source empty")
	}
}

func TestLongestCommonSubstringTargetEmpty(t *testing.T) {
	if longestCommonSubstring("abc", "") != 0 {
		t.Error("expected 0 when target empty")
	}
}

func TestLongestCommonSubstringTwoEmptyStrings(t *testing.T) {
	if longestCommonSubstring("", "") != 0 {
		t.Error("expected 0 for two empty strings")
	}
}

func TestLongestCommonSubstringIdenticalStrings(t *testing.T) {
	if longestCommonSubstring("abc", "abc") != 3 {
		t.Error("expected 3 for identical strings")
	}
}

func TestLongestCommonSubstringCompletelyDifferent(t *testing.T) {
	if longestCommonSubstring("abc", "xyz") != 0 {
		t.Error("expected 0 for completely different strings")
	}
}

func TestLongestCommonSubstringSingleMatchingChar(t *testing.T) {
	if longestCommonSubstring("abc", "xbz") != 1 {
		t.Error("expected 1")
	}
}

func TestLongestCommonSubstringSingleCharMatch(t *testing.T) {
	if longestCommonSubstring("a", "a") != 1 {
		t.Error("expected 1")
	}
}

func TestLongestCommonSubstringSingleCharDiffer(t *testing.T) {
	if longestCommonSubstring("a", "b") != 0 {
		t.Error("expected 0")
	}
}

func TestLongestCommonSubstringAtBeginning(t *testing.T) {
	if longestCommonSubstring("abcdef", "abcxyz") != 3 {
		t.Error("expected 3")
	}
}

func TestLongestCommonSubstringAtEnd(t *testing.T) {
	if longestCommonSubstring("xyzabc", "defabc") != 3 {
		t.Error("expected 3")
	}
}

func TestLongestCommonSubstringPickLongest(t *testing.T) {
	if longestCommonSubstring("abXYZcd", "abXYcd") != 4 {
		t.Error("expected 4")
	}
}

func TestLongestCommonSubstringRepeatedChars(t *testing.T) {
	if longestCommonSubstring("aaaa", "aa") != 2 {
		t.Error("expected 2")
	}
}

File: src/algorithms/strings/edit-distance/longest-common-substring/__tests__/longest-common-substring.test.ts
/** Correctness tests for the longestCommonSubstring pure function. */

import { describe, it, expect } from "vitest";
import { longestCommonSubstring } from "../sources/longest-common-substring.ts?fn";

describe("longestCommonSubstring", () => {
  it('finds the longest common substring between "ABABC" and "BABCBA" (length 4)', () => {
    expect(longestCommonSubstring("ABABC", "BABCBA")).toBe(4);
  });

  it("returns 0 when source is empty", () => {
    expect(longestCommonSubstring("", "abc")).toBe(0);
  });

  it("returns 0 when target is empty", () => {
    expect(longestCommonSubstring("abc", "")).toBe(0);
  });

  it("returns 0 for two empty strings", () => {
    expect(longestCommonSubstring("", "")).toBe(0);
  });

  it("returns full length for identical strings", () => {
    expect(longestCommonSubstring("abc", "abc")).toBe(3);
  });

  it("returns 0 for completely different strings", () => {
    expect(longestCommonSubstring("abc", "xyz")).toBe(0);
  });

  it("finds a single matching character", () => {
    expect(longestCommonSubstring("abc", "xbz")).toBe(1);
  });

  it("handles single-character strings that match", () => {
    expect(longestCommonSubstring("a", "a")).toBe(1);
  });

  it("handles single-character strings that differ", () => {
    expect(longestCommonSubstring("a", "b")).toBe(0);
  });

  it("finds substring at the beginning", () => {
    expect(longestCommonSubstring("abcdef", "abcxyz")).toBe(3);
  });

  it("finds substring at the end", () => {
    expect(longestCommonSubstring("xyzabc", "defabc")).toBe(3);
  });

  it("returns longest when multiple substrings exist", () => {
    expect(longestCommonSubstring("abXYZcd", "abXYcd")).toBe(4);
  });

  it("handles repeated characters", () => {
    expect(longestCommonSubstring("aaaa", "aa")).toBe(2);
  });
});
